DataStructures/BitList.cs

```diff
@@ -0,0 +1,1750 @@
+using System;
+using System.Linq;
+using System.IO;
+using System.Collections;
+using System.Collections.Generic;
+using System.Runtime.InteropServices;
+
+namespace DataStructures
+{
+    /// <summary>
+    /// Exception generated calling a method when the BitList instance is empty.
+    /// </summary>
+    public class EmptyInstanceException: Exception { }
+
+    /// <summary>
+    /// Exception generated when a boolean operation fails.
+    /// </summary>
+    public class BooleanAlgebraException: Exception { }
+
+    /// <summary>
+    /// Exception generated when a BitList instance fails to write or read a file.
+    /// </summary>
+    public class FileException: Exception { }
+
+    /// <summary>
+    /// Exception generated when a BitList instance fails to read a file.
+    /// </summary>
+    public class FileReadingException: FileException { }
+
+    /// <summary>
+    /// Exception generated when a BitList instance fails to write a file.
+    /// </summary>
+    public class FileWritingException: FileException { }
+
+    /// <summary>
+    /// Exception generated when a BitList instance fails the conversion to another type.
+    /// </summary>
+    public class ConversionErrorException: Exception { }
+
+    /// <summary>
+    /// A list of bits.
+    /// </summary>
+    /// <exception cref="EmptyInstanceException" />
+    /// <exception cref="BooleanAlgebraException" />
+    /// <exception cref="FileReadingException" />
+    /// <exception cref="FileWritingException" />
+    /// <exception cref="ConversionErrorException" />
+    public sealed class BitList:
+        IList<bool>,
+        IEquatable<BitList>
+    {
+        private BitArray bits;
+
+        /// <summary>
+        /// No params ctor.
+        /// </summary>
+        public BitList()
+        {
+            bits = null;
+        }
+
+        /// <summary>
+        /// Ctor (Copy the <paramref name="bits"/> <see cref="BitList"/> object).
+        /// </summary>
+        public BitList(BitList bits)
+        {
+            this.bits = new BitArray(bits.bits);
+        }
+
+        /// <summary>
+        /// Ctor (<paramref name="unmanagedNonPrimitiveBits"/> is an unmanaged value type object).
+        /// </summary>
+        public BitList(ValueType unmanagedNonPrimitiveBits)
```

Comment:
Codacy found an issue: [The Cyclomatic Complexity of this constructor is 12 which is greater than 10 authorized.](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=5559704)